program-runtime/src/sysvar_cache.rs

```diff
@@ -23,92 +24,213 @@ impl ::solana_frozen_abi::abi_example::AbiExample for SysvarCache {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct SysvarWithLamports<T: Sysvar> {
+    value: Arc<T>,
+    lamports: u64,
+}
+impl<T: Sysvar> SysvarWithLamports<T> {
+    pub fn new(value: T, lamports: u64) -> Self {
+        Self {
+            value: Arc::new(value),
+            lamports,
+        }
+    }
+    pub fn into_account(&self, rent_epoch: Epoch) -> Account {
+        create_account_with_fields(&*self.value, (self.lamports, rent_epoch))
+    }
+}
+impl<'a, T: Sysvar> TryFrom<&'a AccountSharedData> for SysvarWithLamports<T> {
+    type Error = InstructionError;
+    fn try_from(account: &'a AccountSharedData) -> Result<Self, Self::Error> {
+        let value = bincode::deserialize(account.data())
+            .map_err(|_| InstructionError::UnsupportedSysvar)?;
+        Ok(SysvarWithLamports {
+            value,
+            lamports: account.lamports(),
+        })
+    }
+}
+
 #[derive(Default, Clone, Debug)]
 pub struct SysvarCache {
-    clock: Option<Arc<Clock>>,
-    epoch_schedule: Option<Arc<EpochSchedule>>,
+    /// Used to simulate the rent_epoch on real accounts
+    rent_epoch: Epoch,
```

Comment:
I believe that sysvar accounts have their `rent_epoch` field updated throughout the epoch by eager rent collection. Assuming that's true, each sysvar's rent epoch could be different and setting them all the bank's epoch could cause consensus issues and also make it possible to detect simulation.

I think it would be better to cache the full account info for each sysvar

Reply:
Great catch, I thought this was updated to the bank's epoch during `inherit_specially_retained_account_fields`, but that isn't the case.  I'll add the account in the cache too.